sites/game-client/ui/setting-modal.mjs

```diff
@@ -0,0 +1,291 @@
+      'headphone': 'External Headphone',
+      'bluetooth': 'Bluetooth device',
+    }, 'builtin');
+
+    this.addSwitch('general', 'data_collect', '去識別化資料蒐集', true, 50);
+
+    this.addSection('chat', '聊天設定', 10);
+    this.addDropdown('chat', 'example', '範例下拉選單', 0);
+    this.updateDropdownOptions('chat', 'example', {
+      'default': 'Default',
+      'meow': 'Meow',
+    }, 'default')
+  }
+
+  /**
+   * @param {string} name
+   * @param {string} displayName
+   * @param {number} order
+   */
+  addSection(name, displayName, order) {
+    console.assert(!(name in this.sections), 'section name "' + name + '" is not unique within setting modal.');
+    this.sections[name] = {};
+
+    const elem = document.createElement('div');
+    elem.className = 'setting-section';
+    elem.id = 'setting-section-' + name;
+    elem.style.order = order.toString();
+
+    const title = document.createElement('div');
+    title.className = 'setting-section-title';
+    title.innerText = displayName;
+    elem.append(title);
+    this.mainDOM.append(elem);
+  }
+
+  /**
+   * @param {string} parentSection
+   * @param {string} name - should be unique across all sections and subsections
+   * @param {string} displayName
+   * @param {number} order - order within section
+   */
+  addSubsection(parentSection, name, displayName, order) {
+    console.assert(!(name in this.sections), 'subsection name "' + name + '" is not unique within setting modal.');
+    console.assert(parentSection in this.sections,
+      'subsection "' + name + '" is refer to non-exist section "' + parentSection + '".');
+    this.sections[name] = {};
+
+    const parentElem = document.getElementById('setting-section-' + parentSection);
+    const elem = document.createElement('div');
+    elem.className = 'setting-subsection';
+    elem.id = 'setting-section-' + name;
+    elem.style.order = order.toString();
+
+    const title = document.createElement('div');
+    title.className = 'setting-subsection-title';
+    title.innerText = displayName;
+    elem.append(title);
+    parentElem.append(elem);
+  }
+
+  reset() {
+    for (const section in this.sections) {
+      for (const name in this.sections[section]) {
+        const row = this.sections[section][name];
+        const elem = document.getElementById('setting-item-' + section + '-' + name);
+
+        elem.value = row.currentValue;
+      }
+    }
+  }
+
+  save() {
+    const msg = {};
+
+    for (const section in this.sections) {
+      for (const name in this.sections[section]) {
+        const row = this.sections[section][name];
+        const elem = document.getElementById('setting-item-' + section + '-' + name);
+        if (row.currentValue === elem.value) {
+          continue;
+        }
+
+        const re = new RegExp(row.validation);
+        if (!elem.value.match(re)) {
+          console.warn('Input ' + name + ' doesn\'t match RegEx');
+        }
+        // TODO: send modified setting to the backend
```

Comment:
Not sure how we'll deal with storing settings here for external (extension) user of this class.
We can either:
- Store all settings in one place and allow these external users to retrieve them through a specific API. If we do this, we might need a setting storage class/model.
- Let each individual external users (extension) manage their own settings. If we do this, we might need to allow callers to specify a callback for saving the settings or resetting the settings.

Either way, this is probably something that will come later, and I think we can merge this PR as is.